modules/analysis-runner/src/cljdoc/analysis/runner.clj

```diff
@@ -86,6 +77,34 @@
       (println "Deleting" path)
       (.delete file))))
 
+(defn- prepare-jar-for-analysis!
```

Comment:
I like how you pulled all that io stuff out of `analyze-impl` but I'm not sure if I think that the jar preparation work and class path calculation should be combined like this. An alternative idea: 

- `prepare-jar-for-analysis!` returns `{:jar-contents some-dir :jar-path local-jar-path}`
- `resolved-and-cp` is still called very early in the process (similar place like before)
- this would require `prepare-jar-for-analysis!` to also be called earlier

I think right now we have a call stack that is 

```
analyze-impl -> prepare-jar-for-analysis! -> resolved-and-cp
```
For testing in a REPL and also readability it could be nicer to have something more along the lines of this
```clj
(let [{:keys [jar-contents jar-path]} (prepare-jar-for-analysis! ,,,)
      {:keys [classpath]} (deps/resolved-and-cp jar-path ,,,)]
  (analyze-impl jar-contents classpath ,,,))
```

Let me know what you think :)

Reply:
Hi, Martin!

Now I see that chaining functions like this will impede debugging in REPL. Sure, it's possible to rearrange calls to remove coupling and see the bigger picture. I got more questions, though :)

* I see that downloading and unpacking project JAR over and over again in `analyze-impl` (via `prepare-jar...`) will slow down debugging flow and obstruct the view on the process. But what are benefits of calculating CP separately from jar preparation?
* One point that bugged me in my current design is that `prepare-...` and `analyze-impl` are tightly coupled (`prepare` copies analysis files that will make sense only for `analyze-impl`). That's one of the reasons why call graph looks like this (the other reason is removing duplicate code from `-main` of `runner` and runner-ng`). Do you think it will be better to put copying those files back to `analysis-impl`?
* I got some hard time with temp dir management in this PR... Moving `prepare-jar-...` out of `analyze-impl`, will require some kind of conveying a shared temp dir to both of them. I've already done this for `prepare`, but it still feels kinda clumsy...

Reply:
> But what are benefits of calculating CP separately from jar preparation?

Sometimes when debugging dependency resolution issues I download `.pom` files from Clojars and point the existing code at those files. I imagine something similar might come in handy once this lands. I don't have strong feelings about repeated downloading etc.

> One point that bugged me in my current design is that `prepare-...` and `analyze-impl` are tightly coupled

I'm not sure if there's a solution to this but maybe that's fine? After all the functions depend on another to get the job done.

> I got some hard time with temp dir management in this PR...

I actually missed that fact during my review and that of course makes things a little more complicated. The tmp-dir creation could be moved outside the `analyze-impl` `prepare-` functions as well but then again this adds more arguments to those functions. Could be nice to pass a local directory though, maybe a bit easier to find than system temp dirs.